### ISSUE TITLE
RFC: Structured Tag Generation

### ISSUE THREAD

**Issue by berkoben** [BrickSchema/Brick#36]
**Background:**
The current means of defining Brick tags and Tagsets is highly repetitive, does not enforce consistency / non-duplication, nor does it facilitate generating tag[set] documentation.  

This is a problem when trying to use Brick in real applications.  

**Objective**
To specify a concise method for defining tags and tagsets that allows machine-generation of an internally consistent ontology of tags and tagsets with non-overlapping semantic meaning (i.e. no two tags define exactly the same concept).  

**Discussion:**
One of the most powerful features of Brick is that it provides a means for attaching a precise semantic definition to a piece of data that persists across the context where it is found.  For example, a point tagged  with zone_air_temperature_sensor (ZATS) is known to be the ambient air temperature in the region of space measured by the sensor.  This definition is the same whether the point is obtained from a thermostat or a light fixture (yes, light fixtures sometimes have temp sensors).  

Conceptually, the ZATS tagset is created by combining concepts from disjoint sets to create a more complex concept: <location>_<media>_<measurement type>_<point type>.  This is very powerful because it allows a virtually infinite number of meaningful composite concepts to be created and understood without individually defining each one.  To be more specific about the meaning here:  If each of the individual tags in a tag set has a precise definition, the precise meaning of a tagset will, more often than not, be easy to infer.  (Note: for this to be strictly true all tags must be unique, single words, combinedWithoutSpaces, or make use of a reserved-character-delimiter).

If we are also willing to use the class hierarchies in Brick (pointType, MeasurementType, Equipment, Location, etc...)  to enforce a limited grammar, we could write code to generate internally consistent groups of tagsets very concisely around a central concept.  With the right structure this could be done modularly and flexibly in ways that map to conceptual similarities.  An example: 

Take the concept of temperature.  We could create code that looks something like this: 

```
Temperature:
  group:
    media: [Air]
    types: [Zone,Outside, Discharge, return]
  group:
    media: [Water]
    modifiers: [hot, chilled]
    types: [discharge, return]

would yield: 

Zone_Air_Temperature
Outside_Air_Temperature
Discharge_Air_Temperature
Return_Air_Temperature
Discharge_Hot_Water_Temperature
Discharge_Chilled_Water_Temperature
Return_Hot_Water_Temperature
Return_Chilled_Water_Temperature
```
in the above, we are able to take multiple well-defined tags, and use simple rules to mechanically create tagsets. In the above example, we create groups for air and water because the concepts we want to produce for the two media are divergent.  However, we could easily modify this code if we were monitoring, say an indoor-outdoor pool, to include "water" in the media list of the first group and we would get the following additional tags:

Zone_Water_Temperature
Outside_Water_Temperature
Discharge_Water_Temperature
Return_Water_Temperature

This would allow you to get all of the core concepts written.  A parser for this structured file could ensure that no tagsets appear twice and that only pre-defined tags are used. 

To be added in future posts:  mapping to point types and equipment.

**Comment by gtfierro** [comment]
Just putting the note here that I have an implementation of this that I'm working on and can post soon

**Comment by berkoben** [comment]
Thanks!  I'm writing up comments on your readme right now

**Comment by gtfierro** [comment]
**Note to all**: There is a preliminary implementation of this at the [gabe-efforts branch](https://github.com/BuildSysUniformMetadata/Brick/tree/gabe-efforts/src). It is a work in progress.

First of all, I want to enumerate my support for this idea. Having a structured means of generating tagsets for Brick means that we can be more consistent and complete in its definition, and more easily expand and extend the Brick ontology.  The documents describe the compositions of tags into tagsets, as well as a few of the "meta" relationships between tagsets (such as synonyms).

When these documents are compiled into the Brick ontology, we can make sure that the tagsets are explicitly constructed of the actual tags (e.g. "Outside Air Temperature Sensor" has tags "outside" "air" "temperature" and "sensor").  Keeping these tags consistent has been an issue in the past, but I think its a powerful way of managing the complexity and size of the full class hierarchy. People want to know what they have in their building relating to "discharge air"; the tags can help find what the correct classes to use are.

After giving it some thought, I think there's a couple ways we will want to phrase these documents: a measurement-based view and a class-based view.

## Class-based View

This is a very traditional and straightforward enumeration of a class hierarchy. 

### Overview

Example: 

```yaml
Point:
    subclasses:
        Alarm:
        Command:
        Meter:
        Sensor:
```

or

```yaml
Air Handler Unit:
    synonyms: [AHU]
    subclasses:
        Roof-Top Unit:
            synonyms: [RTU]
```

The advantage here is it is easy to see what the class structure is. As we move down the hierarchy, the equipment becomes more specific in terms of its functionality. There's a role here for something like a `uses` relationship to identify which tagsets are commonly associated with a piece of equipment. For example:

```yaml
Variable Air Volume Box:
    synonyms: [VAV]
    uses:
        - Zone Air Temperature Sensor
        - Discharge Air Temperature Sensor
        - Zone Air Temperature Setpoint
        - Discharge Air Damper
        - Return Air Damper
        - Return Fan
```

I don't quite want to get into the question of "mandatory" vs "optional" attributes yet; this is merely to illustrate the point that being explicit about the class hierarchy dovetails nicely with being able to do this sort of construction.

### Discussion

The problem, as @berkoben has pointed out, is that its difficult to be consistent across sibling branches of the class hierarchy. A concrete example of this is for the concept of temperature above. Within the "Temperature" context, I have both Air temperature and Water temperature. Within each of those, I have discharge and return but also some other types. Further, I have both sensors and setpoints (and differential setpoints, and reset commands, alarms, etc). If we stick with the class-based view, then we're stuck with being explicit about that:

```yaml
Point:
    subclasses:
        Sensor:
            types:
                Temperature:
                    media:
                        - Air:
                               types: [Outside, Return, Discharge]
                        - Water:
                               types: [Return, Discharge]
        Setpoint:
            types:
                Temperature:
                    media:
                        - Air:
                               types: [Discharge]
                        - Water:
                               types: [Discharge]
        Alarm:
            types:
                Temperature:
                    media:
                        - Air:
                               types: [Outside, Return, Discharge]
                        - Water:
                               types: [Return, Discharge]
```

In this case, we have a collection of points (sensors, setpoints, alarms, commands, etc) that are associated with a particular context.

## Measurement-based View

This is what I'm terming the construction that @berkoben offered above. We build the tagsets off of their contexts, which offers a more natural grouping of the points and possibly helps make it more succinct and consistent.

### Overview

```yaml
Temperature:
    - media: Air 
      classes:
          - Sensor:
                 types: [Zone, Outside, Discharge, Return]
          - Setpoint:
                 types: [Zone, Discharge]
    - media: Water
       modifiers: [Hot, Chilled, Condenser]
       types: [Discharge, Return]
       classes: [Sensor, Setpoint]
```

would generate

```
Air Temperature Sensor
Air Temperature Setpoint
Zone Air Temperature Sensor
Outside Air Temperature Sensor
Discharge Air Temperature Sensor
Return Air Temperature Sensor
Zone Air Temperature Setpoint
Discharge Air Temperature Setpoint

Water Temperature Sensor
Water Temperature Setpoint
Hot Water Temperature Sensor
Hot Water Temperature Setpoint
Chilled Water Temperature Sensor
Chilled Water Temperature Setpoint
Condenser Water Temperature Sensor
Condenser Water Temperature Setpoint

Discharge Chilled Water Temperature Sensor
Discharge Chilled Water Temperature Setpoint
Discharge Condenser Water Temperature Sensor
Discharge Condenser Water Temperature Setpoint
Return Chilled Water Temperature Sensor
Return Chilled Water Temperature Setpoint
Return Condenser Water Temperature Sensor
Return Condenser Water Temperature Setpoint
```

### Discussion

I think this makes a lot of sense for the point classes, because they are all grouped around measurements of media (e.g. temperature of water, CO2 content of air, etc). However, this isn't as great for the equipment hierarchy, because a piece of equipment will span several media. For example, a chiller will affect air, but uses water to do so, so it doesn't cleanly fit into either the Air or Water media of a Temperature context.

The disadvantage of this approach is it can start to exhibit some of the "sibling duplication". In my example above,  it doesn't make sense to have Outside and Return air setpoints, so I had to duplicate the `types` field across the classes for the Air Temperature context.

**Comment by berkoben** [comment]
I agree with @gtfierro that it makes sense to construct points and equipment differently.  

As with points, the value in Brick comes from providing a predictable structure while still offering ad-hoc extensibility through extension and composition.  In particular as a programmer, it is important to me that knowing a Brick equipment class for a device tells me something meaningful and specific about it.  As a result, I prefer that Equipment definitions be more granular, numerous and easy to combine (rather than having a smaller number of very broadly defined classes).  The ontology construction tools should make this easy.

Particularly in the case of Equipment, where we deal with real world devices that have endless variation in configuration and features, it's necessary to accept that the coverage of the Brick ontology will always be incomplete.  As such, the real value we provide is the framework for extending what exists to meet a particular user's needs.  

## Detailed Discussion

An important differentiator for Equipment vs. Points is that an Equipment definition is both a concept (A piece of hardware that varies the air volume going into a room) as well as a grouping construct.  For instance a VAV might be defined as having both the above description and the following grouping of points: 

```
Discharge_Air_Flow_Sensor
Discharge_Air_Flow_Setpoint
Damper_Position_Command
```
Java interfaces are a very useful mental model for thinking about Brick equipment.  The functionality is largely analogous, with only small differences depending on whether you think of points as methods or fields http://docs.oracle.com/javase/specs/jls/se7/html/jls-9.html (keep this in mind when we get to code generation!)  

Following this analogy, an interface defines a nugget of functionality ("functionality" being used somewhat loosely here).  This nugget may not be useful or complete on its own, but it is neatly packaged, well described, and can be combined with other nuggets to make complete entities.  These entities are equipment definitions (I leave aside, for the moment, whether Brick should make an explicit distinction between a reusable unit and a "complete" piece of equipment as it is a detail that could be decided later)

In order to construct Equipment we need the ability to: 

- Assign one or more point tagsets to an Equipment class
- Indicate that a particular equipment class extends another equipment class
- Provide a means to specify duplication in the event that a piece of equipment has redundancy (ex: an AHU might have two fans that are identical and interchangeable)

Similarly to points, the **class-based view** as described by @gtfiero in the previous pose causes more problems than it solves here (mainly because multiple inheritance means that we don't have a strict class tree), but the **measurement-based view** can't be applied here either.  Instead, we need a third construct:

## Composition-based view

The underlying ideas behind the composition-based view are: 
- Any composition of points is a meaningful and reusable unit
- The equipment composition is in itself meaningful i.e two Equipment classes that to compose to. 
 i.e. the same set of points could be different if they decompose to non-overlapping equipment superclasses
- The Equipment hierarchy can be defined implicitly and does not have to be specified by the user

Using these ideas a configuration could look something like:

```
VAV:
  Description: A device for....
  Points: [Discharge_Air_Flow_Sensor, Discharge_Air_Flow_Setpoint, Damper_Position_Command]
RH:
  Description: represents points related to controlled air heating using hot water
  Points: [HW_valve_position, 
Discharge_Air_Temp, 
Damper_Sensor]

VAVRH:
  Description: ...
  Subcomponents: [VAV, RH]

CO2:
  Description: ...
  Points: [zone_CO2_sensor, zone_CO2_Setpoint]

VAVRHCO2:
  Description: ...
  Points: [some_point]
  Subcomponents: [VAVRH, CO2]  // equivalent to [VAV, RH, CO2]

```
Duplication could even be built in so that duplicated items get consistent names: 

```
AHU:
  subcomponents[FAN, FAN]
```

Could be used to indicate that there are two identical fans and the generator could name all the associated points `<point>_1` and `<point>_2`, adding them to the point graph as subclasses of the base point 

Additional optimizations could prevent the user from having to independently specify sensor/setpoint/alarm for a particular point:  Ex `zone_temperature_~*!`  Could indicate include all three.  

Finally, the structure above is sufficient for a generator to build all the rdf structure and validate point and equipment names while being very concise.

**Comment by jbkoh** [comment]
Thanks for the detailed implementations and discussions. Though it might have been already late, I add my thoughts.

### Discussion sources
I think all the discussions widely spread, and I think it is best to aggregate all of them here:
- [Keith's private Google doc](https://docs.google.com/document/d/1pcVcpR1p94VF3DGwA5GOkuWXsOHt2JuuvusmS-JadxA)
  (@berkoben, is it possible to make the google doc public for comments?)
- An email thread.
- And Gabe's [README.md](https://github.com/BuildSysUniformMetadata/Brick/tree/gabe-efforts/src)

### About TagSet Composition:

1. We have Resource for media. I think it'd better keep the terminology consistently.

2. I feel that the current design is a bit biased to Point TagSets. If we need semantics of subcomponents of TagSets (for any purpose including TagSet composition), I think each component can be a Tag or another TagSet because a TagSet may include other TagSets like Point having Equipment. And Equipment in point can be also complicated with modifiers like Return_Fan where Return is a modifier or a type or a function. It causes multiple modifiers for different entities inside a TagSet where only Tag dimensions cannot express. An example is "Return_Fan_Differential_Speed_Setpoint" where different modifiers, Return and Differential, specifies different Tags, Fan and Speed.

    (1) It is not a bad idea to avoid such situation by not including Equipment TagSet in Point TagSet, but I can't be sure of whether it is practically right for field engineers.

    (2) Otherwise, I suggest to generalize "TagDimension" to "Dimension" or "InformationDimension" to allow having TagSets.
  
3. Along with 1., I think having complex media (which is Resource in Brick) is better idea than complicating the decision among type, modifier, or function. Then, we can have Return_Air, which is a TagSet, as a subcomponent of Return_Air_Temperature_Sensor.

4. In summary of 1. and 2., **my suggestion** is Point TagSet may have dimensions of Equipment, Resource (or media), and MeasurementType, and each of dimension can have modifiers or functions. Here, it is clearer what each modifier modifies. I think it is also easier to generate TagSets from there. 

5. Is VAVCO2 or VAVRH useful? Querying VAV with Regeat_Coil or VAV with CO2 Sensor may be enough. I think Reheat_VAV could be useful because it has functional meaning. However, I can't agree on usage of CO2_VAV because there is no specification of VAV other than having a CO2 sensor. (If its control loop includes CO2 control, I'd say it is functionally meaningful and useful.)

**Comment by berkoben** [comment]
Private domain docs can't be made globally accessible, however I think Gabe picked up all the comments and integrated them in his long form post, so I believe the RFC thread is comprehensive for the topic of tagset generation.  

Responding to comments by @jbkoh: 

1. Agree.  That was my oversight in using media instead of resource.

2. You bring up a good point that equipment is not included in the point descriptions, however I think the measurement-based view could easily support that use case:
```
DifferentialSpeed: // I encourage using camel case for classes that are compound words
      media: Fan  // note we could change the name from media to something else. See below.
         types: [Discharge, Return]
         classes: [Sensor, Setpoint]

gives you:

Return_Fan_DifferentialSpeed_Sensor
Return_Fan_DifferentialSpeed_Setpoint
Discharge_Fan_DifferentialSpeed_Sensor
Discharge_Fan_DifferentialSpeed_Setpoint
```
In the measurement-based view, differential speed gets its own top-level block (along with temperature, [absolute] speed, etc]).  The name "media" doesn't really make sense for "fan", but "resource" doesn't either, so perhaps "Target"  or something similar? It's also possible to allow "Resource" OR "Equipment", but that level of specification is probably not required (though it would simplify validation).  

3.  see 4.

4.  It appears we are in agreement about:  Media->Resource OR Equipment.  vI think I still prefer the measurementType to be the primary dimension as originally proposed.  

5.  The more time I spend with real HVAC systems, the more I am convinced that complete coverage of real equipment should be a non-goal.  As such I'm less concerned about what Equipment Brick should represent than about what tooling it offers to create specifications for specific equipment beyond the most basic sub-components.

**Comment by berkoben** [comment]
Following up on my post about a **Composition-Based View**, the following is a specification for VAVs from a real building I'm working on:

- **Bold** text is a class definition.  
- Points in the block below the bold text are points that belong to that class.  
- "Extends" notation is used to represent the fact that a particular class inherits the point associations of the extended classes (though I like to think of these things all as interfaces, if we want to be picky).  

In my spec I have multiple named subgroups of points, such as "Occupancy-Status", which don't map to any particular equipment, but are reusable in constructions (and can be defined as representing a particular functionality, if desired).  This approach, which could be used with the construction  tooling I described previously, lets me define four different VAVs with very little configuration. The functional groups can be used across Equipment hierarchies as well, for instance "Occupancy-Status" is usable in a FCU as well.

**VAV-DSP** extends VAV, Zone-Air-Temp-Dual-Setpoint
**VAV-DSP-CO2** extends VAV-DSP, Zone-Air-CO2
**VAV-DSP-RH** extends VAV-DSP, Discharge-Air-Reheat-Valve-Control
**VAV-DSP-RH-CO2** extends VAV-DSP, Discharge-Air-Reheat-Valve-Control, Zone-Air-CO2
 
**VAV** extends Occupancy-Status
Device_Name
Run_Status
Zone_Air_Temperature_Sensor
Discharge_Air_Flow_Sensor
Discharge_Air_Flow_Setpoint
Cooling_Max_Discharge_Air_Flow_Setpoint
Heating_Max_Discharge_Air_Flow_Setpoint
Occupied_Min_Air_Flow_Setpoint
Unoccupied_Min_Air_Flow_Setpoint
Supply_Air_Damper_Position_Command
 
**Occupancy-Status**
Occupancy_Status
Temporary_Occupancy_Status
Calendar_Occupancy_Dtatus
 
**Zone-Air-Temp-Dual-Setpoint**
Cooling_Zone_Air_Temperature_Setpoint
Heating_Zone_Air_Temperature_Setpoint
 
**Zone-Air-CO2**
Zone_CO2_Setpoint
Zone_CO2_Sensor
 
**Discharge-Air-Reheat-Valve-Control**
Heating_Valve_Command
Discharge_Air_Temperature_Sensor
Heating_Max_Discharge_Air_Temperature_Setpoint
Heating_Discharge_Air_Temperature_Setpoint

**Comment by gtfierro** [comment]
I like the idea of using interfaces. The Java-style declaration (e.g. `extends`) feels like a good way to describe what the point/component requirements of that "flavor" of equipment are. What can help with usability is to adopt a structural-typing style approach to determining the instantiated classes. If a VAV `hasPoint` Zone-Air-Temp-Dual-Setpoint, then it *is* a VAV-DSP (per your example above). 

This does not need to be explicitly instantiated; instead it is inferred from the equipment/points attached to an instance of a VAV. This can help manage a large class hiearchy by removing the need for users to read up on what all the classes are when they build their model.

The ontology could link classes to other classes using an `expects` relationship, which would inform the type inference.

```
// type system
VAV     type            Class
VAV     subClassOf      HVAC-Equipment
VAV     expects         Return-Air-Damper
VAV     expects         Supply-Air-Damper
VAV     expects         Zone-Air-Room-Temperature-Setpoint
VAV     expects         Zone-Air-Room-Temperature-Sensor

Reheat-VAV      type    Class
Reheat-VAV      subClassOf  VAV
Reheat-VAV      expects     Heating-Coil

Damper      type        Class
Damper      subClassOf  HVAC-Equipment
Damper      expects     Damper-Position-Sensor
Damper      expects     Damper-Position-Command

Return-Air-Damper    subClassOf      Damper
Supply-Air-Damper    subClassOf      Damper

// instance
VAV123      type        VAV
HC123       type        Heating-Coil
VAV123      hasPart     HC123

// now the following is implicitly true
VAV123      type        Reheat-VAV
```

**Comment by berkoben** [comment]
A couple of thoughts:

@gtfierro What you propose in terms of implicit interface implementations is interesting from the perspective of minimizing the need to know the existing compositions, however it does allow the designer to create excessively verbose configurations (by ignoring the existing compositions).  Additionally, I'm not sure that it is necessarily the case that a composition completely constrains meaning or function.  

To mitigate this, I would propose that the a validation step exists where the designer is prompted to compress redundant field-wise groupings into existing compositions (or mark those field-wise groupings as deliberate).  This would keep the config clean as well as let the system support multiple compositions with the same points but different meanings.  

On the issue of validation, another issue that just came up in our internal discussions is the issue of changing compositions ex-post-facto:.  
 
If a user wants to change the content of an equipment type (a composition of points), it would be useful for the validator to determine the effect of the change on other types.  For Example, let's say we have: 

```
DSP
- point1
- point2
- point3
- point4

VAV 
- point6
- point6

VAV-DSP extends VAV, DSP

and the designer decides to change the composition of DSP

DSP
- point1
- point2

DSP2 
- point3
- point4
- pointOrthogonal

```
The validator should be able to highlight that all the VAV-DSP devices have lost fields such that the the user to edit the schema to ensure that it is backwards compatible:

```
DSP2a
- point3
- point4

VAV-DSP extends VAV, DSP, DSP2a

```

**Comment by bbalaji-ucsd** [comment]
Just reading up on this discussion. I really like the ideas proposed here
and I think it would help “codify” Brick in a much better manner. Did you
guys decide on which way you are planning to go in the final version? Also
wonder how much of this can be made compatible with ontologies so that
SPARQL (and other tools) can work.

There was also a suggestion late last year that we re-use tags from
Haystack to create a compositional model. I think that will also fit in
nicely with the ideas above.

**Comment by berkoben** [comment]
Speaking for myself, I'm pretty happy with the measurement-based view for tagsets and the composition-based view for equipment as described above.  I think this captures a lot of the needs for developers.  

As stated above, I prefer to have explicit rather than implicit inheritance (the only disagreement Gabe and I seem to have above).  I am not familiar enough to sparql to really speak to that aspect, but we can certainly write RDF from the proposed tooling.  

While I like that Haystack tags already have well-formed definitions, I'm not sure that their use of compoundWords as tags fits the Brick composition model very well.  Fro instance we would want "constant" and "volume" to be separate terms in the tooling example above but haystack has constantVolume as a tag.  

I don't know that anything has been undertaken code-wise so far.  Personally I'd prefer to see the focus on the tooling rather than the content. I'd be happy to discuss more specific designs f2f or on VC.

**Comment by bbalaji-ucsd** [comment]
Agree with you.

We could break up the compound tags in Haystack or ignore them to make our
model clear. At this time, the adoption of Haystack is high enough to
warrant use of their vocabulary whenever possible.

**Comment by gtfierro** [comment]
Sorry for being absent in this discussion! I've been busy moving things off my plate so I can get back to this.

I think I'm coming around to @berkoben's view on explicit inheritance. There are some interesting developments in the Haystack working group on their new type system, so I want to take a close look at that to see if there's a way we can bring them together and possibly facilitate translation between them. At the very least we can make sure our definitions jive with theirs.

I haven't done anything code-wise on this. I'm thinking probably the best thing to start with is to generate the Brick RDF from some representation like @berkoben's above for one of the hierarchies. Should be able to do that within the week.

**Comment by gtfierro** [comment]
Given the discussion on #45 , its worth coming back to this discussion so something can be executed and used to more effectively address the other issues that are coming up.

To summarize the above discussion, we've got the following tagset hierarchies and approaches, enumerated in order from most certain to least certain about the design

1. point hierarchy: 
  - Assembling tagsets: use measurement-based approach defined above (still needs refinement). Highest level of the hierarchy is probably the measurement dimension (temperature, pressure, etc) and then we specify point class (sensor, setpoint, ...) and further modifiers.
  - Defining tagsets (**no current solution**): define each of the tags, and then union them for the tagset definition?
2. Equipment hierarchy:
  - Assembling tagsets: use composition-based approach defined above
  - Defining tagsets: could apply definitions to each of the "high-level" classes (e.g. VAV, VAV w/ Reheat) and merge these definitions for each subclass that extends the high-level classes
3. Location hierarchy:
  - Could use composition-based approach, but a simpler, explicit hierarchy composition could be used as well.

I've got a prototype of the point hierarchy that I will fix up and issue as an RFC/PR so we can comment on an actual prototype. Regarding the rest of these, do we want to have an in-the-weeds discussion on a call sometime this week?

@jbkoh @berkoben
